libnetwork: write ServFail if DNS reply msg is bad

corhere opened the pull request:
- Relates to #46928 

If the resolver's DNSBackend returns a name that cannot be marshaled into a well-formed DNS message, the resolver will only discover this when it attempts to write the reply message and it fails with an error. No reply message is sent, leaving the client to wait out its timeout and the user in the dark about what went wrong.

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Make a best-effort attempt to inform the client of the failure.

**- How I did it**

When writing the intended reply message fails, retry once with a ServFail response to inform the client and user that the DNS query was not resolved due to a problem with to the resolver, not the network.

**- How to verify it**

New unit test.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
- The container DNS resolver will send ServFail replies instead of failing silently in more situations.

**- A picture of a cute animal (not mandatory but encouraged)**

corhere committed:
libnetwork: write ServFail if DNS reply msg is bad

If the resolver's DNSBackend returns a name that cannot be marshaled
into a well-formed DNS message, the resolver will only discover this
when it attempts to write the reply message and it fails with an error.
No reply message is sent, leaving the client to wait out its timeout and
the user in the dark about what went wrong.

When writing the intended reply message fails, retry once with a
ServFail response to inform the client and user that the DNS query was
not resolved due to a problem with to the resolver, not the network.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere commented:
cc @robmry

thaJeztah commented on libnetwork/resolver.go:
```diff
@@ -381,6 +381,15 @@ func (r *Resolver) serveDNS(w dns.ResponseWriter, query *dns.Msg) {
+			// has to do with the content of msg rather than the connection.
+			if msg.Rcode != dns.RcodeServerFailure {
+				if err := w.WriteMsg(new(dns.Msg).SetRcode(query, dns.RcodeServerFailure)); err != nil {
+					r.log(ctx).WithError(err).Errorf("[resolver] writing ServFail response also failed")
```
🙈 

```suggestion
					r.log(ctx).WithError(err).Error("[resolver] writing ServFail response also failed")
```

akerouanton reviewed (approved):
Except the broken tests, this LGTM. Thanks!

robmry reviewed (approved):
LGTM too - and the test failure is just duff input ...

```
--- a/libnetwork/resolver_unix_test.go
+++ b/libnetwork/resolver_unix_test.go
@@ -64 +64 @@ func TestDNSIPQuery(t *testing.T) {
-       names := []string{"name1", "NaMe1"}
+       names := []string{"name1.", "NaMe1."}
@@ -87 +87 @@ func TestDNSIPQuery(t *testing.T) {
-       q.SetQuestion("name1", dns.TypeMX)
+       q.SetQuestion("name1.", dns.TypeMX)
@@ -99 +99 @@ func TestDNSIPQuery(t *testing.T) {
-       q.SetQuestion("nonexistent", dns.TypeMX)
+       q.SetQuestion("nonexistent.", dns.TypeMX)
```

corhere committed:
libnetwork: assert DNS replies are well-formed

The well-formedness of a DNS message is only checked when it is
serialized, through the (*dns.Msg).Pack() method. Add a call to Pack()
to our tstwriter mock to mirror the behaviour of the real
dns.ResponseWriter implementation. And fix tests which generated
ill-formed DNS query messages.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere commented:
Should be all fixed now. Serves me right for not running all the package tests locally.

thaJeztah reviewed (approved):
LGTM if CI is ✅

thaJeztah commented:
Windows.. never change. Thanks for the informative errors;

```
=== Failed
=== FAIL: github.com/docker/docker/integration/container TestWindowsDevices/process/class/5B45201D-F2F2-4F3B-85BB-30FF1F953599_mounted (1.48s)
    devices_windows_test.go:107: assertion failed: error is not nil: Error response from daemon: failed to create task for container: failed to create shim task: hcs::CreateComputeSystem 6ca2659c76ba9a58eb85e692ac17b108190d989e5a2f5b23e53d879f2215e8cb: The parameter is incorrect.: unknown
```

```
=== FAIL: github.com/docker/docker/integration/container TestLogs/driver_json-file/tty/stdout_and_stderr (1.50s)
    logs_test.go:134: assertion failed: error is not nil: Error response from daemon: hcs::CreateComputeSystem 202c0491b6a8ccd311fc30340026b2b36165716ae343b77ed7da30103d9008e6: Element not found.
```
